sklearn/cluster/k_means_.py

```diff
@@ -685,7 +685,9 @@ def _init_centroids(X, k, init, random_state=None, x_squared_norms=None,
         seeds = random_state.permutation(n_samples)[:k]
         centers = X[seeds]
     elif hasattr(init, '__array__'):
-        centers = init
+        # ensure that the centers have the same dtype as X
+        # this is a requirement of fused types of cython
+        centers = np.array(init, dtype=X.dtype)
```

Comment:
There is a block of code under `k_means` that starts with `if hasattr(init, '__array__')` that also converts explicitly converts `init` to higher precision.

That should also be fixed and we should test the attributes `cluster_centers_` and `inertia_` for precision...